src/tiled/changemapobject.cpp

```diff
@@ -158,3 +158,35 @@ void ChangeMapObjectsTile::changeTiles()
 
     emit mMapDocument->mapObjectModel()->objectsChanged(mMapObjects);
 }
+
+DetachObject::DetachObject(MapDocument *mapDocument,
+                           MapObject *mapObject)
+    : QUndoCommand(QCoreApplication::translate("Undo Commands",
+                                               "Detach Object",
+                                                nullptr))
+    , mMapDocument(mapDocument)
+    , mMapObject(mapObject)
+{
+    mTemplateRef = mMapObject->templateRef();
+    mProperties = mMapObject->properties();
+}
+
+void DetachObject::redo()
+{
+    // Merge the instance properties into the template properties
+    Properties newProperties = mMapObject->templateObject()->properties();
```

Comment:
This is going to crash if you select both template instances and regular objects and choose "Detach". Either there should be a check here or the `detachObjects` or its caller should filter the objects to make sure they are all template instances (and in fact it should check the return value of `templateObject()` because it may be a broken template reference so `isTemplateInstance` won't be enough).